dart/common/detail/SharedLibraryManager.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2017, Graphics Lab, Georgia Tech Research Corporation
+ * Copyright (c) 2017, Personal Robotics Lab, Carnegie Mellon University
+ * All rights reserved.
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dart/common/detail/SharedLibraryManager.hpp"
+
+#include "dart/common/SharedLibrary.hpp"
+
+namespace dart {
+namespace common {
+namespace detail {
+
+//==============================================================================
+std::shared_ptr<SharedLibrary> SharedLibraryManager::load(
+    const std::string& fileName)
+{
+  const auto iter = mLibraries.find(fileName);
+
+  const auto found = iter != mLibraries.end();
+  if (found)
+  {
+    auto lib = iter->second.lock();
+
+    if (lib)
+      return lib;
+    else
+      mLibraries.erase(iter);
+  }
+
+  const auto newLib = std::make_shared<SharedLibrary>(
+      SharedLibrary::ProtectedConstruction, fileName);
+
+  if (!newLib->isValid())
+    return nullptr;
+
+  mLibraries[fileName] = newLib;
```

Comment:
This performs the same lookup on the map twice. I suggest changing the `find()` above to an `insert()` or `emplace()` and checking the `bool` flag that it returns to check whether the key already exists. That lets you replace this lookup with a simple iterator assignment.

Reply:
Your suggestion is generally the right way to go, but here I try to avoid creating duplicate shared library when it already exists. If I use `emplace()` instead of `find()`, then I have to create the shared library regardless of the existence of the corresponding shared library to the given path.